core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -603,13 +603,21 @@ public static SqlCall stripSeparator(SqlCall call) {
   }
 
   /**
-   * Type-inference strategy for NVL2 function. It returns the least restrictive type
-   * between the second and third operands.
-   */
-  public static final SqlReturnTypeInference NVL2_RESTRICTIVE = opBinding ->
-      opBinding.getTypeFactory().leastRestrictive(
-          Arrays.asList(opBinding.getOperandType(1),
-              opBinding.getOperandType(2)));
+   * Type-inference strategy for NVL2 function.
+   * If the first operand is nullable,
+   * it returns the least restrictive type between the second and third operands.
+   * If the first operand is non-nullable,
+   * it return the second operand type.
+   */
+  public static final SqlReturnTypeInference NVL2_RESTRICTIVE = opBinding -> {
```

Comment:
I remember that this function was adapted by me and it seems to comply with Oracle's specifications.

Reply:
does this mean you think these changes are wrong?

Reply:
`select nvl2(1, 2.0, 3.12);`
In Spark, return 2.00;
In Oracle, return 2;

`select nvl2(1, 2.1, 3.12);`
In Spark, return 2.10;
In Oracle, return 2.1;

Reply:
@NobiGo I think this issue has been discussed in jira
https://issues.apache.org/jira/browse/CALCITE-6397

Reply:
Get it. In Oracle:
`select nvl2(null, 2.1, 3.120);`
return 3.12.
Oracle's return result will automatically omit 0, while Spark does not.

Reply:
Therefore, the current strategy is correct, this PR only modifies whether the value should be nullable.

Reply:
I agree that some exceptions to this are understandable.